java/vector/src/main/java/org/apache/arrow/vector/dictionary/BatchedDictionary.java

```diff
@@ -0,0 +1,312 @@
+
+  private boolean oneTimeEncoding;
+
+  private boolean wasReset;
+
+  private boolean vectorsProvided;
+
+  /**
+   * Creates a dictionary and index vector with the respective types. The dictionary vector
+   * will be named "{name}-dictionary".
+   * <p>
+   * To use this dictionary, provide the dictionary vector to a {@link DictionaryProvider},
+   * add the {@link #getIndexVector()} to the {@link org.apache.arrow.vector.VectorSchemaRoot}
+   * and call the {@link #setSafe(int, byte[], int, int)} or other set methods.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param allocator The allocator to use.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      BufferAllocator allocator
+  ) {
+    this(name, encoding, dictionaryType, indexType, allocator, "-dictionary");
+  }
+
+  /**
+   * Creates a dictionary index vector with the respective types.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param allocator The allocator to use.
+   * @param suffix A non-null suffix to append to the name of the dictionary.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      BufferAllocator allocator,
+      String suffix
+  ) {
+    this(name, encoding, dictionaryType, indexType, allocator, suffix, false);
+  }
+
+  /**
+   * Creates a dictionary index vector with the respective types.
+   *
+   * @param name A name for the vector and dictionary.
+   * @param encoding The dictionary encoding to use.
+   * @param dictionaryType The type of the dictionary data.
+   * @param indexType The type of the encoded dictionary index.
+   * @param allocator The allocator to use.
+   * @param suffix A non-null suffix to append to the name of the dictionary.
+   * @param oneTimeEncoding A mode where the entries can be added to the dictionary until
+   *                        the first stream batch is written. After that, any new entries
+   *                        to the dictionary will throw an exception.
+   */
+  public BatchedDictionary(
+      String name,
+      DictionaryEncoding encoding,
+      ArrowType dictionaryType,
+      ArrowType indexType,
+      BufferAllocator allocator,
+      String suffix,
+      boolean oneTimeEncoding
+  ) {
+    this.encoding = encoding;
+    this.oneTimeEncoding = oneTimeEncoding;
+    if (dictionaryType.getTypeID() != ArrowType.ArrowTypeID.Utf8 &&
+        dictionaryType.getTypeID() != ArrowType.ArrowTypeID.Binary) {
```

Comment:
Can the `dictionaryType.getTypeID()` be `Utf8` and `Binary` at the same time?

Reply:
No, of course, but I was limiting this for string and binary data. Now that I've seen more examples, it looks like folks may use dicts for larger integers or floats as well so a question is whether or not to extend this class for those types or maybe rename this as `BaseBinaryDictionary` to work with only bytes and strings. Thoughts?

Reply:
Dictionary value types are not limited by the spec. I think it's OK to limit support for an initial implementation and file issues (linked to from the source) for expansion.